File: data_baseline/gpt-5/scala/rep3/src/Main.scala
//> using scala "3.3.1"
//> using dep "org.http4s::http4s-ember-server:0.23.34"
//> using dep "org.http4s::http4s-dsl:0.23.34"
//> using dep "org.http4s::http4s-circe:0.23.34"
//> using dep "io.circe::circe-generic:0.14.15"
//> using dep "io.circe::circe-parser:0.14.15"

import cats.effect._
import cats.syntax.all._
import org.http4s._
import org.http4s.dsl.io._
import org.http4s.implicits._
import org.http4s.circe._
import org.http4s.circe.CirceEntityCodec._
import org.http4s.ember.server._
import io.circe._
import io.circe.generic.semiauto._
import io.circe.syntax._
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger
import java.time._
import java.time.temporal.ChronoUnit
import scala.jdk.CollectionConverters._
import com.comcast.ip4s._

object Main extends IOApp {

  case class User(id: Int, username: String)
  case class RegisterRequest(username: String, password: String)
  case class LoginRequest(username: String, password: String)
  case class PasswordChangeRequest(old_password: String, new_password: String)
  case class CreateTodoRequest(title: String, description: Option[String])
  case class UpdateTodoRequest(title: Option[String], description: Option[String], completed: Option[Boolean])

  case class Todo(
      id: Int,
      ownerId: Int,
      title: String,
      description: String,
      completed: Boolean,
      createdAt: String,
      updatedAt: String
  )

  implicit val regReqDecoder: Decoder[RegisterRequest] = deriveDecoder
  implicit val loginReqDecoder: Decoder[LoginRequest] = deriveDecoder
  implicit val pwdChangeDecoder: Decoder[PasswordChangeRequest] = deriveDecoder
  implicit val createTodoDecoder: Decoder[CreateTodoRequest] = deriveDecoder
  implicit val updateTodoDecoder: Decoder[UpdateTodoRequest] = deriveDecoder

  implicit val userEncoder: Encoder[User] = deriveEncoder
  implicit val todoEncoder: Encoder[Todo] = new Encoder[Todo]{
    def apply(t: Todo): Json = Json.obj(
      (
        "id",
        Json.fromInt(t.id)
      ),
      (
        "title",
        Json.fromString(t.title)
      ),
      (
        "description",
        Json.fromString(t.description)
      ),
      (
        "completed",
        Json.fromBoolean(t.completed)
      ),
      (
        "created_at",
        Json.fromString(t.createdAt)
      ),
      (
        "updated_at",
        Json.fromString(t.updatedAt)
      )
    )
  }

  // Storage
  private val usersById = new ConcurrentHashMap[Integer, (String /*username*/, String /*password*/)]()
  private val usersByUsername = new ConcurrentHashMap[String, Integer]()
  private val userIdSeq = new AtomicInteger(0)

  private val sessions = new ConcurrentHashMap[String, Integer]() // token -> userId

  private val todosById = new ConcurrentHashMap[Integer, Todo]()
  private val todoIdSeq = new AtomicInteger(0)

  private def nowInstant(): Instant = Instant.now().truncatedTo(ChronoUnit.SECONDS)
  private def nowIso(): String = nowInstant().toString

  private def errorJson(msg: String): Json = Json.obj("error" -> Json.fromString(msg))

  private def jsonResponse(status: Status, json: Json): IO[Response[IO]] =
    Response[IO](status = status)
      .withEntity(json)
      .withContentType(headers.`Content-Type`(MediaType.application.json))
      .pure[IO]

  private object Auth {
    def extractUserId(req: Request[IO]): Option[(Int, String)] = {
      val maybeCookie = req.cookies.find(_.name == "session_id").map(_.content)
      maybeCookie.flatMap { token =>
        Option(sessions.get(token)).map(i => (i.intValue, token))
      }
    }

    def requireAuth(req: Request[IO])(onAuth: (Int, String) => IO[Response[IO]]): IO[Response[IO]] = {
      extractUserId(req) match {
        case Some((uid, token)) => onAuth(uid, token)
        case None => jsonResponse(Status.Unauthorized, errorJson("Authentication required"))
      }
    }
  }

  private def validateUsername(u: String): Boolean = {
    val lenOk = u.length >= 3 && u.length <= 50
    val regexOk = u.matches("^[A-Za-z0-9_]+$")
    lenOk && regexOk
  }

  private def routes: HttpRoutes[IO] = HttpRoutes.of[IO] {
    // Register
    case req @ POST -> Root / "register" =>
      req.attemptAs[RegisterRequest].value.flatMap {
        case Left(_) => jsonResponse(Status.BadRequest, errorJson("Invalid request"))
        case Right(body) =>
          val username = body.username
          val password = body.password
          if (!validateUsername(username)) {
            jsonResponse(Status.BadRequest, errorJson("Invalid username"))
          } else if (password == null || password.length < 8) {
            jsonResponse(Status.BadRequest, errorJson("Password too short"))
          } else {
            val placeholder = Integer.valueOf(-1)
            val existing = usersByUsername.putIfAbsent(username, placeholder)
            if (existing != null) {
              // username taken
              usersByUsername.replace(username, existing)
              jsonResponse(Status.Conflict, errorJson("Username already exists"))
            } else {
              val id = userIdSeq.incrementAndGet()
              val iid = Integer.valueOf(id)
              usersById.put(iid, (username, password))
              usersByUsername.put(username, iid)
              val user = User(id, username)
              Response[IO](status = Status.Created)
                .withEntity(user)
                .withContentType(headers.`Content-Type`(MediaType.application.json))
                .pure[IO]
            }
          }
      }

    // Login
    case req @ POST -> Root / "login" =>
      req.attemptAs[LoginRequest].value.flatMap {
        case Left(_) => jsonResponse(Status.BadRequest, errorJson("Invalid request"))
        case Right(body) =>
          Option(usersByUsername.get(body.username)) match {
            case None => jsonResponse(Status.Unauthorized, errorJson("Invalid credentials"))
            case Some(iid) =>
              val uid = iid.intValue
              Option(usersById.get(Integer.valueOf(uid))) match {
                case None => jsonResponse(Status.Unauthorized, errorJson("Invalid credentials"))
                case Some(stored) if stored._2 != body.password => jsonResponse(Status.Unauthorized, errorJson("Invalid credentials"))
                case Some(stored) =>
                  val token = java.util.UUID.randomUUID().toString.replaceAll("-", "")
                  sessions.put(token, Integer.valueOf(uid))
                  val user = User(uid, stored._1)
                  val resp = Response[IO](status = Status.Ok)
                    .withEntity(user)
                    .withContentType(headers.`Content-Type`(MediaType.application.json))
                  val cookie = ResponseCookie(name = "session_id", content = token, path = Some("/"), httpOnly = true)
                  IO.pure(resp.addCookie(cookie))
              }
          }
      }

    // Logout
    case req @ POST -> Root / "logout" =>
      Auth.requireAuth(req) { case (_, token) =>
        sessions.remove(token)
        Response[IO](status = Status.Ok)
          .withEntity(Json.obj())
          .withContentType(headers.`Content-Type`(MediaType.application.json))
          .pure[IO]
      }

    // Me
    case req @ GET -> Root / "me" =>
      Auth.requireAuth(req) { case (uid, _) =>
        Option(usersById.get(Integer.valueOf(uid))) match {
          case None => jsonResponse(Status.Unauthorized, errorJson("Authentication required"))
          case Some(userRec) =>
            val user = User(uid, userRec._1)
            Response[IO](status = Status.Ok)
              .withEntity(user)
              .withContentType(headers.`Content-Type`(MediaType.application.json))
              .pure[IO]
        }
      }

    // Password change
    case req @ PUT -> Root / "password" =>
      Auth.requireAuth(req) { case (uid, _) =>
        req.attemptAs[PasswordChangeRequest].value.flatMap {
          case Left(_) => jsonResponse(Status.BadRequest, errorJson("Invalid request"))
          case Right(body) =>
            Option(usersById.get(Integer.valueOf(uid))) match {
              case None => jsonResponse(Status.Unauthorized, errorJson("Authentication required"))
              case Some(userRec) if userRec._2 != body.old_password => jsonResponse(Status.Unauthorized, errorJson("Invalid credentials"))
              case Some(userRec) =>
                if (body.new_password == null || body.new_password.length < 8) jsonResponse(Status.BadRequest, errorJson("Password too short"))
                else {
                  usersById.put(Integer.valueOf(uid), (userRec._1, body.new_password))
                  Response[IO](status = Status.Ok)
                    .withEntity(Json.obj())
                    .withContentType(headers.`Content-Type`(MediaType.application.json))
                    .pure[IO]
                }
            }
        }
      }

    // List todos
    case req @ GET -> Root / "todos" =>
      Auth.requireAuth(req) { case (uid, _) =>
        val list = todosById.asScala.values.filter(_.ownerId == uid).toList.sortBy(_.id)
        Response[IO](status = Status.Ok)
          .withEntity(list)
          .withContentType(headers.`Content-Type`(MediaType.application.json))
          .pure[IO]
      }

    // Create todo
    case req @ POST -> Root / "todos" =>
      Auth.requireAuth(req) { case (uid, _) =>
        req.attemptAs[CreateTodoRequest].value.flatMap {
          case Left(_) => jsonResponse(Status.BadRequest, errorJson("Invalid request"))
          case Right(body) =>
            val title = Option(body.title).getOrElse("")
            if (title.trim.isEmpty) jsonResponse(Status.BadRequest, errorJson("Title is required"))
            else {
              val id = todoIdSeq.incrementAndGet()
              val created = nowIso()
              val todo = Todo(
                id = id,
                ownerId = uid,
                title = title,
                description = body.description.getOrElse(""),
                completed = false,
                createdAt = created,
                updatedAt = created
              )
              todosById.put(Integer.valueOf(id), todo)
              Response[IO](status = Status.Created)
                .withEntity(todo)
                .withContentType(headers.`Content-Type`(MediaType.application.json))
                .pure[IO]
            }
        }
      }

    // Get todo by id
    case req @ GET -> Root / "todos" / IntVar(id) =>
      Auth.requireAuth(req) { case (uid, _) =>
        Option(todosById.get(Integer.valueOf(id))) match {
          case None => jsonResponse(Status.NotFound, errorJson("Todo not found"))
          case Some(todo) if todo.ownerId != uid => jsonResponse(Status.NotFound, errorJson("Todo not found"))
          case Some(todo) =>
            Response[IO](status = Status.Ok)
              .withEntity(todo)
              .withContentType(headers.`Content-Type`(MediaType.application.json))
              .pure[IO]
        }
      }

    // Update todo by id (partial)
    case req @ PUT -> Root / "todos" / IntVar(id) =>
      Auth.requireAuth(req) { case (uid, _) =>
        Option(todosById.get(Integer.valueOf(id))) match {
          case None => jsonResponse(Status.NotFound, errorJson("Todo not found"))
          case Some(todo) if todo.ownerId != uid => jsonResponse(Status.NotFound, errorJson("Todo not found"))
          case Some(todo) =>
            req.attemptAs[UpdateTodoRequest].value.flatMap {
              case Left(_) => jsonResponse(Status.BadRequest, errorJson("Invalid request"))
              case Right(body) =>
                body.title match {
                  case Some(t) if t.trim.isEmpty => jsonResponse(Status.BadRequest, errorJson("Title is required"))
                  case _ =>
                    val newTitle = body.title.getOrElse(todo.title)
                    val newDesc = body.description.getOrElse(todo.description)
                    val newCompleted = body.completed.getOrElse(todo.completed)
                    val now = nowInstant()
                    val prev = Instant.parse(todo.updatedAt)
                    val chosen = if (now.isAfter(prev)) now else prev.plusSeconds(1)
                    val updated = todo.copy(
                      title = newTitle,
                      description = newDesc,
                      completed = newCompleted,
                      updatedAt = chosen.truncatedTo(ChronoUnit.SECONDS).toString
                    )
                    todosById.put(Integer.valueOf(id), updated)
                    Response[IO](status = Status.Ok)
                      .withEntity(updated)
                      .withContentType(headers.`Content-Type`(MediaType.application.json))
                      .pure[IO]
                }
            }
        }
      }

    // Delete todo by id
    case req @ DELETE -> Root / "todos" / IntVar(id) =>
      Auth.requireAuth(req) { case (uid, _) =>
        Option(todosById.get(Integer.valueOf(id))) match {
          case None => jsonResponse(Status.NotFound, errorJson("Todo not found"))
          case Some(todo) if todo.ownerId != uid => jsonResponse(Status.NotFound, errorJson("Todo not found"))
          case Some(_) =>
            todosById.remove(Integer.valueOf(id))
            IO.pure(Response[IO](status = Status.NoContent))
        }
      }
  }

  def program(port: Int): IO[Unit] = {
    val httpApp = routes.orNotFound

    val serverResource = EmberServerBuilder.default[IO]
      .withHost(host"0.0.0.0")
      .withPort(Port.fromInt(port).get)
      .withHttpApp(httpApp)
      .build

    serverResource.use { _ =>
      IO.never
    }
  }

  private def parseArgs(args: List[String]): Int = {
    // default port 8080; look for --port PORT
    def loop(rest: List[String], port: Int): Int = rest match {
      case Nil => port
      case "--port" :: p :: tail => loop(tail, p.toIntOption.getOrElse(port))
      case _ :: tail => loop(tail, port)
    }
    loop(args, 8080)
  }

  override def run(args: List[String]): IO[ExitCode] = {
    val port = parseArgs(args)
    program(port).as(ExitCode.Success)
  }
}

File: data_baseline/gpt-5/java/rep3/src/Main.java
import java.io.*;
import java.net.*;
import com.sun.net.httpserver.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.time.*;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;

public class Main {
    public static void main(String[] args) throws Exception {
        int port = 8080;
        for (int i = 0; i < args.length; i++) {
            if ("--port".equals(args[i]) && i + 1 < args.length) {
                port = Integer.parseInt(args[i + 1]);
            }
        }
        InetSocketAddress addr = new InetSocketAddress("0.0.0.0", port);
        HttpServer server = HttpServer.create(addr, 0);
        TodoApp app = new TodoApp();
        server.createContext("/", app::handle);
        server.setExecutor(Executors.newCachedThreadPool());
        server.start();
        System.out.println("Server started on 0.0.0.0:" + port);
    }
}

class TodoApp {
    private final Map<Integer, User> users = new ConcurrentHashMap<>();
    private final Map<String, Integer> usernameToId = new ConcurrentHashMap<>();
    private final Map<String, Integer> sessions = new ConcurrentHashMap<>(); // token -> userId
    private final Map<Integer, Todo> todos = new ConcurrentHashMap<>();
    private final Map<Integer, List<Integer>> userTodos = new ConcurrentHashMap<>(); // userId -> list of todo ids (ordered)
    private final AtomicInteger userIdSeq = new AtomicInteger(1);
    private final AtomicInteger todoIdSeq = new AtomicInteger(1);

    public void handle(HttpExchange ex) throws IOException {
        try {
            String method = ex.getRequestMethod();
            URI uri = ex.getRequestURI();
            String path = uri.getPath();
            if (path == null || path.isEmpty()) path = "/";

            // Routing
            if (method.equals("POST") && path.equals("/register")) {
                handleRegister(ex);
            } else if (method.equals("POST") && path.equals("/login")) {
                handleLogin(ex);
            } else if (method.equals("POST") && path.equals("/logout")) {
                withAuth(ex, (user) -> handleLogout(ex, user));
            } else if (method.equals("GET") && path.equals("/me")) {
                withAuth(ex, (user) -> handleMe(ex, user));
            } else if (method.equals("PUT") && path.equals("/password")) {
                withAuth(ex, (user) -> handlePassword(ex, user));
            } else if (path.equals("/todos") && method.equals("GET")) {
                withAuth(ex, (user) -> handleListTodos(ex, user));
            } else if (path.equals("/todos") && method.equals("POST")) {
                withAuth(ex, (user) -> handleCreateTodo(ex, user));
            } else if (path.startsWith("/todos/") && path.length() > 7) {
                String idStr = path.substring(7);
                Integer id = parseId(idStr);
                if (id == null) {
                    sendJson(ex, 404, errorJson("Todo not found"));
                    return;
                }
                if (method.equals("GET")) {
                    withAuth(ex, (user) -> handleGetTodo(ex, user, id));
                } else if (method.equals("PUT")) {
                    withAuth(ex, (user) -> handleUpdateTodo(ex, user, id));
                } else if (method.equals("DELETE")) {
                    withAuth(ex, (user) -> handleDeleteTodo(ex, user, id));
                } else {
                    sendJson(ex, 404, errorJson("Not found"));
                }
            } else {
                sendJson(ex, 404, errorJson("Not found"));
            }
        } catch (Exception e) {
            // Log and return 500
            e.printStackTrace();
            sendJson(ex, 500, errorJson("Internal server error"));
        } finally {
            try { ex.close(); } catch (Exception ignore) {}
        }
    }

    private static Integer parseId(String s) {
        try { return Integer.parseInt(s); } catch (Exception e) { return null; }
    }

    private interface AuthedHandler { void run(User user) throws IOException; }

    private void withAuth(HttpExchange ex, AuthedHandler handler) throws IOException {
        String token = getSessionToken(ex.getRequestHeaders());
        if (token == null) {
            sendJson(ex, 401, errorJson("Authentication required"));
            return;
        }
        Integer uid = sessions.get(token);
        if (uid == null) {
            sendJson(ex, 401, errorJson("Authentication required"));
            return;
        }
        User user = users.get(uid);
        if (user == null) {
            sendJson(ex, 401, errorJson("Authentication required"));
            return;
        }
        handler.run(user);
    }

    private String getSessionToken(Headers headers) {
        List<String> list = headers.get("Cookie");
        if (list == null) return null;
        for (String cookieHeader : list) {
            String[] parts = cookieHeader.split(";\\s*");
            for (String part : parts) {
                int eq = part.indexOf('=');
                if (eq > 0) {
                    String name = part.substring(0, eq).trim();
                    String val = part.substring(eq + 1).trim();
                    if (name.equals("session_id")) {
                        return val;
                    }
                }
            }
        }
        return null;
    }

    // Endpoint handlers
    private void handleRegister(HttpExchange ex) throws IOException {
        String body = readBody(ex);
        Map<String, Object> json = Json.parseObject(body);
        String username = str(json.get("username"));
        String password = str(json.get("password"));
        if (username == null || !username.matches("^[a-zA-Z0-9_]{3,50}$")) {
            sendJson(ex, 400, errorJson("Invalid username"));
            return;
        }
        if (password == null || password.length() < 8) {
            sendJson(ex, 400, errorJson("Password too short"));
            return;
        }
        synchronized (this) {
            if (usernameToId.containsKey(username)) {
                sendJson(ex, 409, errorJson("Username already exists"));
                return;
            }
            int id = userIdSeq.getAndIncrement();
            User user = new User(id, username, password);
            users.put(id, user);
            usernameToId.put(username, id);
            sendJson(ex, 201, user.toPublicJson());
        }
    }

    private void handleLogin(HttpExchange ex) throws IOException {
        String body = readBody(ex);
        Map<String, Object> json = Json.parseObject(body);
        String username = str(json.get("username"));
        String password = str(json.get("password"));
        if (username == null || password == null) {
            sendJson(ex, 401, errorJson("Invalid credentials"));
            return;
        }
        Integer id = usernameToId.get(username);
        if (id == null) {
            sendJson(ex, 401, errorJson("Invalid credentials"));
            return;
        }
        User user = users.get(id);
        if (user == null || !user.password.equals(password)) {
            sendJson(ex, 401, errorJson("Invalid credentials"));
            return;
        }
        String token = UUID.randomUUID().toString().replaceAll("-", "");
        sessions.put(token, user.id);
        Headers resp = ex.getResponseHeaders();
        resp.add("Set-Cookie", "session_id=" + token + "; Path=/; HttpOnly");
        sendJson(ex, 200, user.toPublicJson());
    }

    private void handleLogout(HttpExchange ex, User user) throws IOException {
        String token = getSessionToken(ex.getRequestHeaders());
        if (token != null) {
            sessions.remove(token);
        }
        sendJson(ex, 200, "{}");
    }

    private void handleMe(HttpExchange ex, User user) throws IOException {
        sendJson(ex, 200, user.toPublicJson());
    }

    private void handlePassword(HttpExchange ex, User user) throws IOException {
        String body = readBody(ex);
        Map<String, Object> json = Json.parseObject(body);
        String oldp = str(json.get("old_password"));
        String newp = str(json.get("new_password"));
        if (oldp == null || !user.password.equals(oldp)) {
            sendJson(ex, 401, errorJson("Invalid credentials"));
            return;
        }
        if (newp == null || newp.length() < 8) {
            sendJson(ex, 400, errorJson("Password too short"));
            return;
        }
        user.password = newp;
        sendJson(ex, 200, "{}");
    }

    private void handleListTodos(HttpExchange ex, User user) throws IOException {
        List<Integer> ids = userTodos.getOrDefault(user.id, Collections.emptyList());
        // Ensure ascending order by id
        List<Integer> copy = new ArrayList<>(ids);
        Collections.sort(copy);
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        boolean first = true;
        for (Integer id : copy) {
            Todo t = todos.get(id);
            if (t == null) continue;
            if (!first) sb.append(",");
            sb.append(t.toJson());
            first = false;
        }
        sb.append("]");
        sendJson(ex, 200, sb.toString());
    }

    private void handleCreateTodo(HttpExchange ex, User user) throws IOException {
        String body = readBody(ex);
        Map<String, Object> json = Json.parseObject(body);
        String title = str(json.get("title"));
        String description = str(json.get("description"));
        if (title == null || title.trim().isEmpty()) {
            sendJson(ex, 400, errorJson("Title is required"));
            return;
        }
        if (description == null) description = "";
        int id = todoIdSeq.getAndIncrement();
        String now = nowIso();
        Todo t = new Todo(id, user.id, title, description, false, now, now);
        todos.put(id, t);
        userTodos.computeIfAbsent(user.id, k -> Collections.synchronizedList(new ArrayList<>())).add(id);
        sendJson(ex, 201, t.toJson());
    }

    private void handleGetTodo(HttpExchange ex, User user, int id) throws IOException {
        Todo t = todos.get(id);
        if (t == null || t.userId != user.id) {
            sendJson(ex, 404, errorJson("Todo not found"));
            return;
        }
        sendJson(ex, 200, t.toJson());
    }

    private void handleUpdateTodo(HttpExchange ex, User user, int id) throws IOException {
        Todo t = todos.get(id);
        if (t == null || t.userId != user.id) {
            sendJson(ex, 404, errorJson("Todo not found"));
            return;
        }
        String body = readBody(ex);
        Map<String, Object> json = Json.parseObject(body);
        boolean any = false;
        if (json.containsKey("title")) {
            String title = str(json.get("title"));
            if (title == null || title.trim().isEmpty()) {
                sendJson(ex, 400, errorJson("Title is required"));
                return;
            }
            t.title = title;
            any = true;
        }
        if (json.containsKey("description")) {
            String description = str(json.get("description"));
            if (description == null) description = "";
            t.description = description;
            any = true;
        }
        if (json.containsKey("completed")) {
            Boolean c = bool(json.get("completed"));
            if (c != null) {
                t.completed = c.booleanValue();
                any = true;
            }
        }
        if (any) {
            t.updatedAt = nowIso();
        }
        sendJson(ex, 200, t.toJson());
    }

    private void handleDeleteTodo(HttpExchange ex, User user, int id) throws IOException {
        Todo t = todos.get(id);
        if (t == null || t.userId != user.id) {
            sendEmpty(ex, 404, errorJson("Todo not found"));
            return;
        }
        todos.remove(id);
        List<Integer> ids = userTodos.get(user.id);
        if (ids != null) ids.remove((Integer)id);
        // DELETE success: 204 no body
        sendNoContent(ex);
    }

    private static String str(Object o) { return (o instanceof String) ? (String)o : null; }
    private static Boolean bool(Object o) { return (o instanceof Boolean) ? (Boolean)o : null; }

    private static String errorJson(String msg) { return "{\"error\":\"" + Json.escape(msg) + "\"}"; }

    private static String nowIso() {
        return DateTimeFormatter.ISO_INSTANT.format(Instant.now().truncatedTo(ChronoUnit.SECONDS));
    }

    private static String readBody(HttpExchange ex) throws IOException {
        InputStream is = ex.getRequestBody();
        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        byte[] buf = new byte[4096];
        int n;
        while ((n = is.read(buf)) != -1) bos.write(buf, 0, n);
        String charset = "UTF-8";
        return bos.toString(charset);
    }

    private static void sendJson(HttpExchange ex, int status, String json) throws IOException {
        Headers headers = ex.getResponseHeaders();
        headers.set("Content-Type", "application/json");
        byte[] bytes = json.getBytes("UTF-8");
        ex.sendResponseHeaders(status, bytes.length);
        OutputStream os = ex.getResponseBody();
        os.write(bytes);
        os.flush();
    }

    private static void sendEmpty(HttpExchange ex, int status, String json) throws IOException {
        // For error on DELETE we still return JSON body
        sendJson(ex, status, json);
    }

    private static void sendNoContent(HttpExchange ex) throws IOException {
        // 204 no body
        ex.getResponseHeaders().set("Content-Type", "application/json");
        ex.sendResponseHeaders(204, -1);
    }
}

class User {
    public final int id;
    public final String username;
    public String password; // stored in plain for simplicity (in-memory only)

    public User(int id, String username, String password) {
        this.id = id;
        this.username = username;
        this.password = password;
    }

    public String toPublicJson() {
        return "{\"id\":" + id + ",\"username\":\"" + Json.escape(username) + "\"}";
    }
}

class Todo {
    public final int id;
    public final int userId;
    public String title;
    public String description;
    public boolean completed;
    public String createdAt;
    public String updatedAt;

    public Todo(int id, int userId, String title, String description, boolean completed, String createdAt, String updatedAt) {
        this.id = id;
        this.userId = userId;
        this.title = title;
        this.description = description;
        this.completed = completed;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
    }

    public String toJson() {
        StringBuilder sb = new StringBuilder();
        sb.append("{");
        sb.append("\"id\":").append(id).append(",");
        sb.append("\"title\":\"").append(Json.escape(title)).append("\",");
        sb.append("\"description\":\"").append(Json.escape(description)).append("\",");
        sb.append("\"completed\":").append(completed).append(",");
        sb.append("\"created_at\":\"").append(Json.escape(createdAt)).append("\",");
        sb.append("\"updated_at\":\"").append(Json.escape(updatedAt)).append("\"");
        sb.append("}");
        return sb.toString();
    }
}

class Json {
    // Minimal JSON utilities for our use-case
    public static Map<String,Object> parseObject(String s) throws IOException {
        if (s == null) return new HashMap<>();
        Parser p = new Parser(s);
        Object v = p.parseValue();
        if (!(v instanceof Map)) return new HashMap<>();
        return (Map<String,Object>)v;
    }

    public static String escape(String s) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"': sb.append("\\\""); break;
                case '\\': sb.append("\\\\"); break;
                case '\n': sb.append("\\n"); break;
                case '\r': sb.append("\\r"); break;
                case '\t': sb.append("\\t"); break;
                default:
                    if (c < 0x20) {
                        sb.append(String.format("\\u%04x", (int)c));
                    } else {
                        sb.append(c);
                    }
            }
        }
        return sb.toString();
    }

    static class Parser {
        private final String s;
        private int i = 0;
        Parser(String s) { this.s = s; }

        private void skipWs() {
            while (i < s.length()) {
                char c = s.charAt(i);
                if (c == ' ' || c == '\n' || c == '\r' || c == '\t') i++; else break;
            }
        }

        Object parseValue() throws IOException {
            skipWs();
            if (i >= s.length()) return null;
            char c = s.charAt(i);
            if (c == '{') return parseObject();
            if (c == '"') return parseString();
            if (c == 't' || c == 'f') return parseBoolean();
            if (c == 'n') { parseNull(); return null; }
            // number not needed but parse minimal
            if (c == '-' || (c >= '0' && c <= '9')) return parseNumber();
            return null;
        }

        Map<String,Object> parseObject() throws IOException {
            Map<String,Object> map = new LinkedHashMap<>();
            expect('{');
            skipWs();
            if (peek('}')) { i++; return map; }
            while (true) {
                skipWs();
                String key = parseString();
                skipWs();
                expect(':');
                Object val = parseValue();
                map.put(key, val);
                skipWs();
                if (peek(',')) { i++; continue; }
                if (peek('}')) { i++; break; }
                throw new IOException("Invalid JSON object");
            }
            return map;
        }

        String parseString() throws IOException {
            expect('"');
            StringBuilder sb = new StringBuilder();
            while (i < s.length()) {
                char c = s.charAt(i++);
                if (c == '"') break;
                if (c == '\\') {
                    if (i >= s.length()) throw new IOException("Invalid escape");
                    char e = s.charAt(i++);
                    switch (e) {
                        case '"': sb.append('"'); break;
                        case '\\': sb.append('\\'); break;
                        case '/': sb.append('/'); break;
                        case 'b': sb.append('\b'); break;
                        case 'f': sb.append('\f'); break;
                        case 'n': sb.append('\n'); break;
                        case 'r': sb.append('\r'); break;
                        case 't': sb.append('\t'); break;
                        case 'u':
                            if (i + 4 > s.length()) throw new IOException("Invalid unicode escape");
                            String hex = s.substring(i, i+4);
                            i += 4;
                            sb.append((char)Integer.parseInt(hex, 16));
                            break;
                        default:
                            throw new IOException("Invalid escape");
                    }
                } else {
                    sb.append(c);
                }
            }
            return sb.toString();
        }

        Boolean parseBoolean() throws IOException {
            if (s.startsWith("true", i)) { i += 4; return Boolean.TRUE; }
            if (s.startsWith("false", i)) { i += 5; return Boolean.FALSE; }
            throw new IOException("Invalid boolean");
        }

        void parseNull() throws IOException {
            if (s.startsWith("null", i)) { i += 4; return; }
            throw new IOException("Invalid null");
        }

        Number parseNumber() throws IOException {
            int start = i;
            if (s.charAt(i) == '-') i++;
            while (i < s.length() && Character.isDigit(s.charAt(i))) i++;
            if (i < s.length() && s.charAt(i) == '.') {
                i++;
                while (i < s.length() && Character.isDigit(s.charAt(i))) i++;
            }
            String num = s.substring(start, i);
            try {
                if (num.contains(".")) return Double.parseDouble(num);
                else return Long.parseLong(num);
            } catch (NumberFormatException e) {
                throw new IOException("Invalid number");
            }
        }

        private void expect(char ch) throws IOException {
            skipWs();
            if (i >= s.length() || s.charAt(i) != ch) throw new IOException("Expected '" + ch + "'");
            i++;
        }
        private boolean peek(char ch) {
            skipWs();
            return i < s.length() && s.charAt(i) == ch;
        }
    }
}
